src/pinecone.ts

```diff
@@ -134,12 +134,23 @@ export class Pinecone {
     this._createIndex = createIndex(api);
     this._describeCollection = describeCollection(api);
     this._deleteCollection = deleteCollection(api);
-    this._describeIndex = describeIndex(api);
+    this._describeIndex = describeIndex(api, this._handleDescribeIndex);
     this._deleteIndex = deleteIndex(api);
     this._listCollections = listCollections(api);
     this._listIndexes = listIndexes(api);
   }
 
+  /**
+   * @internal
+   * Used as a callback by {@link Pinecone.describeIndex} to cache any index host URLs that are returned.
+   * This bypasses the need to fetch the host on subsequent dataplane calls.
+   */
+  _handleDescribeIndex = (response: IndexDescription, indexName: IndexName) => {
+    if (response.status?.host) {
+      HostUrlSingleton._set(this.config, indexName, response.status.host);
+    }
+  };
+
```

Comment:
Good point, not really sure why I decided to go with a callback at this point. This would have been much cleaner. 😂

Reply:
I do remember the reason: we're returning promises and I went with passing an explicit callback rather than using `.then()` on the return. Using the promise itself and keeping things isolated to `pinecone.ts`.